http2: add altsvc support

jasnell opened the pull request:
Add support for sending and receiving ALTSVC frames.

As defined by https://tools.ietf.org/html/rfc7838

ping @nodejs/http2
##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
http2

TimothyGu commented on src/node_http2.cc:
```diff
@@ -2379,6 +2416,42 @@ void Http2Stream::RefreshState(const FunctionCallbackInfo<Value>& args) {
+  int32_t id = args[0]->Int32Value(env->context()).ToChecked();
+
+  // TODO(jasnell): This is not exactly right as these are supposed to
+  // be ASCII only.
```
Is this TODO not currently solvable? We know for a fact that at least  the origin is ASCII-only.

jasnell replied:
Yep, was already working on it :-)

TimothyGu commented on doc/api/http2.md:
```diff
@@ -558,11 +558,65 @@ added: REPLACEME
+* `stream` {number} The numeric identifier of a stream (or `0`). Defaults to
+  `0`.
+
+Submits an `ALTSVC` frame (as defined by [RFC 7838][]) to the connected client.
```
Would help if what stream ID 0 (and why it must have an origin) could be explained. Not everyone likes to read RFC’s.

jasnell commented on lib/internal/http2/core.js:
```diff
@@ -1154,6 +1165,40 @@ class ServerHttp2Session extends Http2Session {
+      throw new errors.RangeError('ERR_OUT_OF_RANGE', 'stream');
+
+    if (origin)
+      origin = (new URL(origin)).origin;
```
There is a check for a falsy origin on line 1195.

TimothyGu replied:
A “null origin” is not that the returned origin is `null` but that it is `'null'`. Try `new URL('abc:').origin`, which IMO should be forbidden.

jasnell replied:
ah, right, forgot about that little nuance. Sigh.

jasnell replied:
Ok, added a check for `'null'` in the code and in the test

jasnell commented on src/node_http2.cc:
```diff
@@ -1080,6 +1089,34 @@ inline void Http2Session::HandleGoawayFrame(const nghttp2_frame* frame) {
+    Integer::New(isolate, id),
+    String::NewFromOneByte(isolate,
+                           altsvc->origin,
+                           v8::NewStringType::kInternalized,
```
nah, cut'n'paste error actually

jasnell commented:
@addaleax ... renamed the public API function as suggested.

TimothyGu commented on doc/api/http2.md:
```diff
@@ -558,11 +558,70 @@ added: REPLACEME
 added: v8.4.0
 -->
 
+#### Event: 'altsvc'
```
I sort of feel this name should correspond with the name of the method on the server side. I’m ambivalent about whether alternateService is a better name than altsvc though: I feel eventually most people would refer to it as ALTSVC (HTTP/2 frame) or Alt-Svc (HTTP/1.1 header), while technically the name of the header is alternateService*s*.

jasnell replied:
@addaleax ... given that you were the one that made the original rename suggestion, what do you think here?

addaleax replied:
@jasnell If everybody here agrees that it’s going to be referred to as `ALTSVC` anyway, yea, probably best to stick with it?

Either way, yes, +1 for being consistent :)

jasnell replied:
I think `altsvc` is definitely going to be the most common given the way the header and frame type are defined. I'll switch the API call back :-)

TimothyGu commented on src/node_http2.cc:
```diff
@@ -2379,6 +2416,44 @@ void Http2Stream::RefreshState(const FunctionCallbackInfo<Value>& args) {
+
+  int32_t id = args[0]->Int32Value(env->context()).ToChecked();
+
+  // origin and value are both required to be ASCII, handle them as such.
```
Is this requirement enforced for the value string? If not then we should, to combat the HTTP header encoding debacle that only recently got rectified. In fact IMO we should verify that it only consists of HTTP tokens, similar to the HTTP header checks. 

Origin is fine because it’s guaranteed to be ASCII by the URL parser.

jasnell replied:
The value is not limited to `token`. It will contain a sequence of `uri-host` constructs (`host:port`) within quoted strings. At the C++ level, we write these out as one-byte strings so there is at least some level of enforcement there. And on the *receiving* side, these are treated as one byte strings so the value will be received correctly. Anyone using extended characters will just see those get corrupted. Because of the performance implications and the fact that this is a non-critical API, I don't think we need to validate every character.... although, it would be good to extend the documentation on this a bit more -- as you said, not everyone likes to read RFCs :-)

TimothyGu replied:
The additional documentation is always a plus. 

> Anyone using extended characters will just see those get corrupted.

Well that’s certainly not the best of outcomes. If this is truly a non-critical API, I think we should ensure correctness above performance. Good call on quotation marks though, and in that case I think simply using a `/[\x00-\x7f]/` check should be enough.

TimothyGu replied:
There may also be some security implications  involved, when `'\uff20'` is treated identically to `'\u0020' by the V8 API.

jasnell replied:
Ok, added a check

jasnell commented on doc/api/http2.md:
```diff
@@ -558,11 +558,70 @@ added: REPLACEME
+Sending an `ALTSVC` frame with stream ID `0` (the default) indicates that
+the alternate service is associated with the given `origin`. Sending an
+`ALTSVC` frame with a specific stream ID indicates that the alternate
+service is associated with the origin of the given `Http2Stream`.
```
I've been stewing over it. I'm starting to have a real dislike for polymorphic function signatures but it would make things a bit easier in both the implementation and on the user.

jasnell commented on lib/internal/http2/core.js:
```diff
@@ -1154,6 +1165,40 @@ class ServerHttp2Session extends Http2Session {
+    if (len > 16382)  // Max length permitted for ALTSVC
+      throw new errors.TypeError('ERR_HTTP2_ALTSVC_LENGTH');
+
+    if (stream === 0 && !origin || origin.length === 0 || origin === 'null')
```
If we simplify the signature to allow the second argument to be either a string or a number, then we don't necessarily have to jump through this hoop. Will stew on it a bit more

jasnell commented:
@TimothyGu @addaleax ... updated! Went with the two argument signature and reworked the error codes a bit. Also extended the documentation to be more useful. PTAL

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/12359/

jasnell commented on doc/api/http2.md:
```diff
@@ -558,11 +558,95 @@ added: REPLACEME
+
+The format of the `alt` parameter is strictly defined by [RFC 7838][] as an
+ASCII string containing a comma-delimited list of "alternative" protocols
+associated with a specific host and port.
```
Yeah, good idea

jasnell commented on lib/internal/http2/core.js:
```diff
@@ -32,6 +32,8 @@ const kMaxFrameSize = (2 ** 24) - 1;
 const kMaxInt = (2 ** 32) - 1;
 const kMaxStreams = (2 ** 31) - 1;
 
+const kQuotedString = /^[\x09\x32\x21-\x5b\x5d-\x7e\x80-\xff]*$/;
```
oh ha!! thinking hex, typed decimal. sigh

jasnell replied:
I'd still prefer to keep the x80-xff range given the definition of quoted-string. Yes, it *shouldn't* happen, but the spec definition technically allows for it. It will be up to the user to validate that what they get is a valid value.

TimothyGu replied:
Okay. I can’t agree with that interpretation of the spec but this is not a blocker for me like the corruption issue was.

TimothyGu commented on lib/internal/http2/core.js:
```diff
@@ -1154,6 +1168,42 @@ class ServerHttp2Session extends Http2Session {
+
+    if (typeof originOrStream === 'string') {
+      origin = (new URL(originOrStream)).origin;
+      if (origin.length === 0 || origin === 'null')
```
The length check is unnecessary: serialized origins cannot be empty.

mcollina commented on doc/api/http2.md:
```diff
@@ -558,11 +558,97 @@ added: REPLACEME
+* `alt` {string} A description of the alternative service configuration as
+  defined by [RFC 7838][].
+* `originOrStream` {string|number} Either a URL string specifying the origin or
+  the numeric identifier of an active `Http2Stream`.
```
it is not clear where this number can be found.

mcollina commented on lib/internal/http2/core.js:
```diff
@@ -1154,6 +1168,42 @@ class ServerHttp2Session extends Http2Session {
+    let origin;
+
+    if (typeof originOrStream === 'string') {
+      origin = (new URL(originOrStream)).origin;
```
I think we would be better of in accepting a `URL` object rather than a string, or an object with an `origin`  property. The typical usage is to send this down for every session, and passing an object allows a user to avoid parsing a URL for every request.

jasnell replied:
Let's support both. If `originOrStream` is passed in as an object with an `origin` property, it will be used. This risks, of course, the value not being strictly ASCII, but we can document that requirement.

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/12379/

jasnell commented:
CI is good. One failure on freebsd is unrelated.

jasnell commented:
@mcollina ... I plan to get this landed in the morning. Can you take one final look?

jasnell commented:
Landed in ce22d6f
